File: utils/udp_audio_client.py
#!/usr/bin/env python3
"""
UDP Audio Client - Receives audio from Raspberry Pi Pico via UDP
Saves audio to WAV file and displays statistics
"""

import socket
import struct
import wave
import time
import argparse
import sys

def main():
    parser = argparse.ArgumentParser(description='UDP Audio Client - Receive audio from RP2350')
    parser.add_argument('--host', default='192.168.12.200', help='Pico IP address')
    parser.add_argument('--port', type=int, default=5001, help='UDP audio port')
    parser.add_argument('--output', default='audio.wav', help='Output WAV file')
    parser.add_argument('--duration', type=int, default=15, help='Capture duration in seconds')
    parser.add_argument('--sample-rate', type=int, default=48000, help='Expected sample rate (Hz)')
    parser.add_argument('--frame-size', type=int, default=528, help='Samples per frame')
    
    args = parser.parse_args()
    
    print(f"[INFO] UDP Audio Client")
    print(f"[INFO] Target: {args.host}:{args.port}")
    print(f"[INFO] Expected sample rate: {args.sample_rate} Hz")
    print(f"[INFO] Frame size: {args.frame_size} samples")
    print(f"[INFO] Output: {args.output}")
    print(f"[INFO] Duration: {args.duration} seconds")
    print(f"[INFO] Press Ctrl+C to stop capture\n")
    
    # Create UDP socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_RCVBUF, 1024 * 1024)  # 1MB receive buffer
    sock.settimeout(0.5)  # Shorter timeout for better responsiveness
    
    # Send initial packet to register with server
    try:
        sock.sendto(b'HELLO', (args.host, args.port))
        print(f"[INFO] Sent registration packet to {args.host}:{args.port}")
    except Exception as e:
        print(f"[ERROR] Failed to send registration: {e}")
        return 1
    
    # Open WAV file for writing
    wav_file = wave.open(args.output, 'wb')
    wav_file.setnchannels(1)  # Mono
    wav_file.setsampwidth(2)  # 16-bit
    wav_file.setframerate(args.sample_rate)
    
    # Statistics
    start_time = time.time()
    total_samples = 0
    total_frames = 0
    last_stats_time = start_time
    frame_size_bytes = args.frame_size * 2  # 16-bit samples
    lost_frames = 0
    last_seq = None
    timeout_count = 0
    max_timeouts = 0
    
    try:
        print(f"[INFO] Listening for audio packets (Frame size: {frame_size_bytes} bytes)...")
        while True:
            # Check if we've exceeded duration
            elapsed = time.time() - start_time
            if elapsed > args.duration:
                print(f"\n[INFO] Duration limit reached ({args.duration}s)")
                break
            
            try:
                # Receive UDP packet with timeout
                data, (src_ip, src_port) = sock.recvfrom(65536)
                timeout_count = 0  # Reset timeout counter on successful receive
                
                if len(data) >= frame_size_bytes:
                    # Extract audio samples (16-bit signed, little-endian)
                    samples = struct.unpack(f'<{args.frame_size}h', data[:frame_size_bytes])
                    
                    # Write to WAV file
                    wav_file.writeframes(struct.pack(f'<{args.frame_size}h', *samples))
                    
                    # **CRITICAL FIX**: Flush the underlying file buffer after each write
                    # This prevents buffer overflow issues at ~12 seconds
                    # wave.Wave_write uses _file internally, so flush that
                    try:
                        if hasattr(wav_file, '_file'):
                            wav_file._file.flush()
                    except (AttributeError, ValueError):
                        pass
                    
                    total_samples += args.frame_size
                    total_frames += 1
                    
                    # Print stats every second
                    now = time.time()
                    if now - last_stats_time >= 1.0:
                        elapsed_seconds = now - start_time
                        sample_rate = total_samples / elapsed_seconds if elapsed_seconds > 0 else 0
                        print(f"[CAPTURE] {total_samples} samples ({total_samples * 2} bytes), "
                              f"frames: {total_frames}, rate: {sample_rate:.0f} Hz ({elapsed_seconds:.1f}s)")
                        last_stats_time = now
                        max_timeouts = max(max_timeouts, timeout_count)
                else:
                    # Frame too small, skip with warning
                    print(f"[WARNING] Received {len(data)} bytes, expected at least {frame_size_bytes} bytes")
            
            except socket.timeout:
                # No data available within timeout period
                timeout_count += 1
                
                # Alert if timeouts are happening frequently (indicates network issues)
                if timeout_count >= 3:
                    elapsed = time.time() - start_time
                    print(f"[WARNING] No data for {timeout_count * 0.5:.1f}s at elapsed {elapsed:.1f}s - "
                          f"Network issue or Pico not sending? "
                          f"(Received {total_frames} frames so far)")
                    timeout_count = 0  # Reset to avoid spam
                
                pass  # Continue waiting
    
    except KeyboardInterrupt:
        print(f"\n[INFO] Capture stopped by user")
    
    except Exception as e:
        print(f"[ERROR] Receive error: {e}")
        import traceback
        traceback.print_exc()
    
    finally:
        # Flush and close file to ensure all data is written
        try:
            if hasattr(wav_file, '_file'):
                wav_file._file.flush()
        except (AttributeError, ValueError):
            pass
        wav_file.close()
        sock.close()
        
        # Final statistics
        elapsed = time.time() - start_time
        if elapsed > 0:
            avg_rate = total_samples / elapsed
            print(f"\n[INFO] Capture complete!")
            print(f"[INFO] Total samples: {total_samples}")
            print(f"[INFO] Total frames: {total_frames}")
            print(f"[INFO] Average rate: {avg_rate:.0f} Hz")
            print(f"[INFO] Elapsed time: {elapsed:.2f}s")
            print(f"[INFO] File: {args.output}")
            print(f"[INFO] Lost frames: {lost_frames}")
        
        return 0

if __name__ == '__main__':
    sys.exit(main())

File: utils/udp_audio_client.cpp
/**
 * UDP Audio Client for VibeCode4
 *
 * Receives audio from Raspberry Pi Pico via UDP and saves to raw audio file or analyzes tone.
 * Handles sequence numbers and reports packet loss statistics.
 *
 * Usage:
 *     ./udp_audio_client --host 192.168.12.200 --port 5001 --duration 30 --output audio.raw
 *     ./udp_audio_client --host 192.168.12.200 --port 5001 --duration 10 --analyze-tone
 *
 * Compile: g++ -std=c++17 -O2 -lm -o udp_audio_client udp_audio_client.cpp
 */

#include <iostream>
#include <vector>
#include <cstring>
#include <cstdint>
#include <cmath>
#include <chrono>
#include <iomanip>
#include <fstream>
#include <algorithm>

#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <signal.h>
#include <cerrno>

static bool should_exit = false;

void signal_handler(int) {
    should_exit = true;
}

/**
 * Extract the sequence number from the first 4 bytes (little-endian)
 */
uint32_t extract_sequence(const uint8_t* data, size_t len) {
    if (len < 4) {
        return 0;
    }
    return (static_cast<uint32_t>(data[0]) |
            (static_cast<uint32_t>(data[1]) << 8) |
            (static_cast<uint32_t>(data[2]) << 16) |
            (static_cast<uint32_t>(data[3]) << 24));
}

/**
 * Analyze audio buffer for tone frequency (simple energy in frequency bands)
 */
void analyze_tone(const int16_t* samples, int count, int sample_rate) {
    // Simple tone detection: compute RMS and report
    int64_t sum_sq = 0;
    for (int i = 0; i < count; i++) {
        sum_sq += (int64_t)samples[i] * samples[i];
    }
    double rms = std::sqrt((double)sum_sq / count);
    
    // Peak detection
    int16_t peak = 0;
    for (int i = 0; i < count; i++) {
        if (std::abs(samples[i]) > peak) {
            peak = std::abs(samples[i]);
        }
    }
    
    std::cout << "  [AUDIO] RMS: " << std::fixed << std::setprecision(1) 
              << rms << ", Peak: " << peak << std::endl;
}

void receive_audio(const std::string& host, int port, int duration, 
                   const std::string& output_file, bool analyze_tone_flag) {
    std::cout << "[INFO] UDP Audio Client" << std::endl;
    std::cout << "[INFO] Target: " << host << ":" << port << std::endl;
    std::cout << "[INFO] Duration: " << duration << " seconds" << std::endl;
    if (!output_file.empty()) {
        std::cout << "[INFO] Output file: " << output_file << std::endl;
    }
    if (analyze_tone_flag) {
        std::cout << "[INFO] Analyzing tone characteristics" << std::endl;
    }
    std::cout << std::endl;

    // Create UDP socket
    int sock = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock < 0) {
        std::cerr << "[ERROR] Failed to create socket: " << strerror(errno) << std::endl;
        return;
    }

    // Set socket options
    int reuse = 1;
    setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse));

    // Increase receive buffer
    int rcvbuf_size = 16 * 1024 * 1024;
    setsockopt(sock, SOL_SOCKET, SO_RCVBUF, &rcvbuf_size, sizeof(rcvbuf_size));

    // Bind to listen on any interface
    struct sockaddr_in addr;
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_ANY);
    addr.sin_port = htons(port);

    if (bind(sock, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        std::cerr << "[ERROR] Failed to bind socket: " << strerror(errno) << std::endl;
        close(sock);
        return;
    }

    // Set socket timeout for responsive exit
    struct timeval tv;
    tv.tv_sec = 1;
    tv.tv_usec = 0;
    setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));

    // Send registration packet
    struct sockaddr_in pico_addr;
    pico_addr.sin_family = AF_INET;
    pico_addr.sin_port = htons(port);
    inet_pton(AF_INET, host.c_str(), &pico_addr.sin_addr);

    if (sendto(sock, "HELLO", 5, 0, (struct sockaddr*)&pico_addr, sizeof(pico_addr)) < 0) {
        std::cerr << "[WARNING] Failed to send registration packet: " << strerror(errno) << std::endl;
    } else {
        std::cout << "[INFO] Registration packet sent to " << host << ":" << port << std::endl;
    }

    // Setup signal handler
    signal(SIGINT, signal_handler);

    // Open output file if specified
    std::ofstream outfile;
    if (!output_file.empty()) {
        outfile.open(output_file, std::ios::binary);
        if (!outfile) {
            std::cerr << "[ERROR] Failed to open output file: " << output_file << std::endl;
            close(sock);
            return;
        }
    }

    // Statistics
    auto start_time = std::chrono::steady_clock::now();
    uint32_t total_frames = 0;
    uint32_t lost_frames = 0;
    uint32_t last_seq = 0;
    bool first_packet = true;
    auto last_stats_time = start_time;
    
    const int FRAME_SIZE = 528;  // samples per frame
    const int FRAME_BYTES = FRAME_SIZE * 2;

    std::cout << "[INFO] Listening for audio packets..." << std::endl << std::endl;

    uint8_t buffer[65536];
    struct sockaddr_in src_addr;
    socklen_t src_addr_len = sizeof(src_addr);

    while (!should_exit) {
        // Check duration
        auto now = std::chrono::steady_clock::now();
        double elapsed_sec = std::chrono::duration<double>(now - start_time).count();
        if (elapsed_sec > duration) {
            std::cout << "[INFO] Duration limit reached" << std::endl;
            break;
        }

        // Receive packet
        int n = recvfrom(sock, buffer, sizeof(buffer), 0, 
                        (struct sockaddr*)&src_addr, &src_addr_len);
        
        if (n < 0) {
            // Timeout - just continue
            continue;
        }

        if (n < 4 + FRAME_BYTES) {
            continue;  // Invalid packet size
        }

        // Extract sequence and audio
        uint32_t seq = extract_sequence(buffer, n);
        int16_t* audio_samples = (int16_t*)(buffer + 4);

        // Detect frame loss
        if (!first_packet) {
            uint32_t gap = seq - last_seq - 1;
            if (gap > 0) {
                lost_frames += gap;
                std::cout << "[WARNING] Lost " << gap << " frames (seq " 
                         << last_seq << " -> " << seq << ")" << std::endl;
            }
        }

        first_packet = false;
        last_seq = seq;
        total_frames++;

        // Write audio data to file
        if (outfile) {
            outfile.write((char*)audio_samples, FRAME_BYTES);
        }

        // Analyze tone if requested
        if (analyze_tone_flag && total_frames % 10 == 0) {
            analyze_tone(audio_samples, FRAME_SIZE, 48000);
        }

        // Print statistics every second
        now = std::chrono::steady_clock::now();
        if (std::chrono::duration<double>(now - last_stats_time).count() >= 1.0) {
            double total_frames_expected = (lost_frames + total_frames);
            double loss_pct = total_frames_expected > 0 ? 
                (100.0 * lost_frames / total_frames_expected) : 0.0;
            
            std::cout << "[STATS] Frames: " << total_frames 
                     << ", Lost: " << lost_frames 
                     << ", Loss: " << std::fixed << std::setprecision(2) << loss_pct << "%"
                     << ", Elapsed: " << std::setprecision(1) << elapsed_sec << "s" << std::endl;
            
            last_stats_time = now;
        }
    }

    // Final statistics
    std::cout << std::endl << "======================================" << std::endl;
    auto total_time = std::chrono::steady_clock::now() - start_time;
    double elapsed = std::chrono::duration<double>(total_time).count();
    
    double total_frames_expected = lost_frames + total_frames;
    double loss_pct = total_frames_expected > 0 ? 
        (100.0 * lost_frames / total_frames_expected) : 0.0;
    
    std::cout << "[FINAL STATS]" << std::endl;
    std::cout << "  Duration: " << std::fixed << std::setprecision(2) << elapsed << " seconds" << std::endl;
    std::cout << "  Frames received: " << total_frames << std::endl;
    std::cout << "  Frames lost: " << lost_frames << std::endl;
    std::cout << "  Total frames: " << (uint32_t)total_frames_expected << std::endl;
    std::cout << "  Loss rate: " << std::setprecision(2) << loss_pct << "%" << std::endl;
    
    if (!output_file.empty()) {
        std::cout << "  Audio saved to: " << output_file << std::endl;
    }

    if (outfile) {
        outfile.close();
    }

    close(sock);
}

int main(int argc, char* argv[]) {
    std::string host = "192.168.12.200";
    int port = 5001;
    int duration = 15;
    std::string output_file;
    bool analyze_tone_flag = false;

    // Parse arguments
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--host" && i + 1 < argc) {
            host = argv[++i];
        } else if (arg == "--port" && i + 1 < argc) {
            port = std::stoi(argv[++i]);
        } else if (arg == "--duration" && i + 1 < argc) {
            duration = std::stoi(argv[++i]);
        } else if (arg == "--output" && i + 1 < argc) {
            output_file = argv[++i];
        } else if (arg == "--analyze-tone") {
            analyze_tone_flag = true;
        } else if (arg == "--help") {
            std::cout << "Usage: " << argv[0] << " [options]" << std::endl;
            std::cout << "  --host <ip>         Pico IP address (default: 192.168.12.200)" << std::endl;
            std::cout << "  --port <port>       UDP port (default: 5001)" << std::endl;
            std::cout << "  --duration <sec>    Capture duration in seconds (default: 15)" << std::endl;
            std::cout << "  --output <file>     Save audio as raw PCM (default: none)" << std::endl;
            std::cout << "  --analyze-tone      Analyze tone characteristics" << std::endl;
            return 0;
        }
    }

    receive_audio(host, port, duration, output_file, analyze_tone_flag);
    return 0;
}
